File: mamda/c_cpp/src/gunittest/trades/MamdaTradeListenerShortSaleTests.cpp
/* $Id$
 *
 * OpenMAMA: The open middleware agnostic messaging API
 * Copyright (C) 2011 NYSE Technologies, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */

#include <gtest/gtest.h>

#include <mama/MamaDictionary.h>
#include <mama/mamacpp.h>
#include <mama/MamaMsg.h>

#include <mamda/MamdaTradeFields.h>
#include <mamda/MamdaCommonFields.h>
#include <mamda/MamdaSubscription.h>
#include <mamda/MamdaConfig.h>
#include <mamda/MamdaTradeListener.h>
#include <mamda/MamdaTradeHandler.h>
#include <mamda/MamdaDataException.h>
#include <mamda/MamdaFieldState.h>
#include <mamda/MamdaErrorListener.h>
#include <mamda/MamdaQualityListener.h>

#include "common/MamdaUnitTestUtils.h"
#include "common/CpuTestGenerator.h"
#include "common/MemoryTestGenerator.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <cstdlib>
#include <sstream>
#include <iostream>
#include <fstream>

using std::string;

using namespace std;
using namespace Wombat;

class MamdaTradeListenerShortSaleTests : public ::testing::Test
{
protected:
    MamdaTradeListenerShortSaleTests () {}
    virtual ~MamdaTradeListenerShortSaleTests () {}  

    virtual void SetUp()
    {
        try
        {
            mamaBridge bridge;
            bridge = Mama::loadBridge("wmw");
            Mama::open();
            //mama_enableLogging (stderr, MAMA_LOG_LEVEL_FINEST);
            mDictionary = new MamaDictionary;
            if (!mDictionary)
            {
                FAIL() << "Failed to allocate MamaDictionary\n";
                return;
            }
            
            mDictionary->populateFromFile("dictionary.txt");
            MamdaCommonFields::setDictionary (*mDictionary);
            MamdaTradeFields::reset();
            MamdaTradeFields::setDictionary (*mDictionary);
                      
            mSubscription = new MamdaSubscription;
            if (!mSubscription)
            {
                FAIL() << "Failed to allocate MamdaSubscription\n";
                return;
            }
            mTradeListener = new MamdaTradeListener; 
            if (!mTradeListener)
            {
                FAIL() << "Failed to allocate MamdaTradeListener\n";
                return;
            }
            mSubscription->addMsgListener(mTradeListener);
          
        }
        catch(MamaStatus status)
        {
            FAIL() << "Failed to setup Mamda: "
            << status.toString() << "\n";
            return;
        }
    }    
    
    virtual void TearDown()
    {     
        if (mDictionary)
        {     
            delete mDictionary;
            mDictionary = NULL;
        }  
        if (mTradeListener)
        {
            delete mTradeListener;
            mTradeListener = NULL;
        }        
    }

    void callMamdaOnMsg(MamdaSubscription* sub, MamaMsg& msg)
    {  
        try
        {   
            vector<MamdaMsgListener*>& msgListeners = sub->getMsgListeners();
            unsigned long size = msgListeners.size();
            for (unsigned long i = 0; i < size; i++)           
            {        
                MamdaMsgListener* listener = msgListeners[i];
                listener->onMsg (sub, msg,(short)msg.getType());
            }   
        }                
        catch (std::exception &e)
        {
            std::cerr << "Exception: " << e.what() << std::endl;
        }
        catch (...)
        {
            std::cerr << "Exception: Unknown" << std::endl;
        }
        return;
    }
  
    MamaDictionary*         mDictionary;
    MamdaTradeListener*     mTradeListener;
    MamdaSubscription*      mSubscription;
};

TEST_F(MamdaTradeListenerShortSaleTests, tradelistenerRecapFuncTest)
{
    MamaMsg* mMsg = NULL;
    mMsg = new MamaMsg;
    mMsg->create();
    addMamaHeaderFields(*mMsg,
                        MAMA_MSG_TYPE_INITIAL,
                        MAMA_MSG_STATUS_OK,
                        1);
                        
    mMsg->addChar("wShortSaleCircuitBreaker",5248,'A');    
    callMamdaOnMsg(mSubscription,*mMsg);  
    EXPECT_TRUE('A' == mTradeListener->getShortSaleCircuitBreaker());
    EXPECT_EQ(MODIFIED,mTradeListener->getShortSaleCircuitBreakerFieldState());   
}

TEST_F(MamdaTradeListenerShortSaleTests, tradelistenerTradeFuncTest)
{
    MamaMsg* mMsg = NULL;
    mMsg = new MamaMsg;
    mMsg->create();
    addMamaHeaderFields(*mMsg,
                        MAMA_MSG_TYPE_TRADE,
                        MAMA_MSG_STATUS_OK,
                        1);                        
                        
    mMsg->addChar("wShortSaleCircuitBreaker",5248,'A');    
    callMamdaOnMsg(mSubscription,*mMsg);  
    EXPECT_TRUE('A' == mTradeListener->getShortSaleCircuitBreaker());
    EXPECT_EQ(MODIFIED,mTradeListener->getShortSaleCircuitBreakerFieldState());          
   
}

TEST_F(MamdaTradeListenerShortSaleTests, tradelistenerTradeCorrectionFuncTest)
{
    MamaMsg* mMsg = NULL;
    mMsg = new MamaMsg;
    mMsg->create();
    addMamaHeaderFields(*mMsg,
                        MAMA_MSG_TYPE_CORRECTION,
                        MAMA_MSG_STATUS_OK,
                        1);
                        
    mMsg->addChar("wCorrShortSaleCircuitBreaker",5250,'A');    
    callMamdaOnMsg(mSubscription,*mMsg);  
    EXPECT_TRUE('A' == mTradeListener->getCorrShortSaleCircuitBreaker());
    EXPECT_EQ(MODIFIED,mTradeListener->getCorrShortSaleCircuitBreakerFieldState());           
    
}

TEST_F(MamdaTradeListenerShortSaleTests, tradelistenerTradeCanelFuncTest)
{
    MamaMsg* mMsg = NULL;
    mMsg = new MamaMsg;
    mMsg->create();
    addMamaHeaderFields(*mMsg,
                        MAMA_MSG_TYPE_CORRECTION,
                        MAMA_MSG_STATUS_OK,
                        1);
                        
    mMsg->addChar("wOrigShortSaleCircuitBreaker",5249,'A');    
    callMamdaOnMsg(mSubscription,*mMsg);  
    EXPECT_TRUE('A' == mTradeListener->getOrigShortSaleCircuitBreaker());
    EXPECT_EQ(MODIFIED,mTradeListener->getOrigShortSaleCircuitBreakerFieldState());           
    
}

File: mamda/dotnet/src/NUnittest/trades/MamdaTradeListenerShortSaleTests.cs
using System;
using System.Collections.Generic;
using System.Collections;
using NUnit.Framework;
using Wombat;

namespace NUnitTest
{
    [TestFixture]
    public class MamdaTradeListenerShortSaleTests
    {
        public MamdaSubscription mSubscription = null;
        public MamdaTradeListener mTradeListener = null;
        public MamaMsg mMsg = null;
        public MamaBridge myBridge = null;
        public MamaDictionary mDictionary = null;
        public tradetickerCB ticker = null;

        [SetUp]
        public void SetUp()
        {
            try
            {
                //Intialise mama
                myBridge = new MamaBridge("lbm");
                Mama.open();

                mDictionary = new MamaDictionary();
                mDictionary.create("dictionary.txt");

                MamdaCommonFields.setDictionary(mDictionary, null);
                MamdaTradeFields.reset();
                MamdaTradeFields.setDictionary(mDictionary, null);

                mSubscription = new MamdaSubscription();
                mTradeListener = new MamdaTradeListener();
                mSubscription.addMsgListener(mTradeListener);

                ticker = new tradetickerCB();
                mTradeListener.addHandler(ticker);
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e.ToString());
                Environment.Exit(1);
            }
        }

        [Test]
        public void testShortSaleCircuitBreaker()
        {
            mMsg = new MamaMsg();
            addMamaHeaderFields(mMsg,
                                (byte)mamaMsgType.MAMA_MSG_TYPE_TRADE,
                                (byte)mamaMsgStatus.MAMA_MSG_STATUS_OK,
                                (uint)1);
            mMsg.addChar("wShortSaleCircuitBreaker", 5248, 'C');
            ticker.callMamdaOnMsg(mSubscription, mMsg);
            Assert.AreEqual('C', ticker.myShortSaleCircuitBreaker);
		}

        [Test]
        public void testOrigShortSaleCircuitBreaker()
        {
            mMsg = new MamaMsg();
            addMamaHeaderFields(mMsg,
                                (byte)mamaMsgType.MAMA_MSG_TYPE_CANCEL,
                                (byte)mamaMsgStatus.MAMA_MSG_STATUS_OK,
                                (uint)1);
            mMsg.addChar("wOrigShortSaleCircuitBreaker", 5249, 'C');
            ticker.callMamdaOnMsg(mSubscription, mMsg);
            Assert.AreEqual('C', ticker.myOrigShortSaleCircuitBreaker);
        }

        [Test]
        public void testCorrShortSaleCircuitBreaker()
        {
            mMsg = new MamaMsg();
            addMamaHeaderFields(mMsg,
                                (byte)mamaMsgType.MAMA_MSG_TYPE_CORRECTION,
                                (byte)mamaMsgStatus.MAMA_MSG_STATUS_OK,
                                (uint)1);
            mMsg.addChar("wCorrShortSaleCircuitBreaker", 5250, 'C');
            ticker.callMamdaOnMsg(mSubscription, mMsg);
            Assert.AreEqual('C', ticker.myCorrShortSaleCircuitBreaker);
        }
           
        public void addMamaHeaderFields(
                            MamaMsg msg,
                            byte msgType,
                            byte msgStatus,
                            uint seqNum)
        {
            msg.addU8(null, 1, msgType);
            msg.addU8(null, 2, msgStatus);
            msg.addU32(null, 10, seqNum);
            msg.addU64(null, 20, 1);
        }

        public class tradetickerCB : MamdaTradeHandler,
                                     MamdaErrorListener,
                                     MamdaStaleListener
        {
            public char myShortSaleCircuitBreaker;
            public char myOrigShortSaleCircuitBreaker;
            public char myCorrShortSaleCircuitBreaker;

            public void callMamdaOnMsg(MamdaSubscription sub, MamaMsg msg)
            {
                try
                {
                        ArrayList msgListeners = new ArrayList();
                        msgListeners = sub.getMsgListeners();
                        IEnumerator iter = msgListeners.GetEnumerator();
                        int size = msgListeners.Count;
                        for (int i = 0; i < size; i++)
                        {
                            iter.MoveNext();
                            MamdaMsgListener listener = (MamdaMsgListener)iter.Current;

                            listener.onMsg(sub, msg, msg.getType());
                        }
                    }
					catch (Exception ex)
                {
                        throw new MamdaDataException(ex.Message, ex);
                    }
            }
			
			public void onTradeRecap(
                MamdaSubscription sub,
                MamdaTradeListener listener,
                MamaMsg msg,
                MamdaTradeRecap recap)
            {
            }

            public void onTradeReport(
                MamdaSubscription sub,
                MamdaTradeListener listener,
                MamaMsg msg,
                MamdaTradeReport trade,
                    MamdaTradeRecap recap)
            {
                myShortSaleCircuitBreaker = trade.getShortSaleCircuitBreaker();
                Console.WriteLine("myShortSaleCircuitBreaker: " + myShortSaleCircuitBreaker);
            }

            public void onTradeGap(
                MamdaSubscription sub,
                MamdaTradeListener listener,
                MamaMsg msg,
                MamdaTradeGap gap,
                MamdaTradeRecap recap)
            {
            }

            public void onTradeCancelOrError(
                MamdaSubscription sub,
                MamdaTradeListener listener,
                MamaMsg msg,
                MamdaTradeCancelOrError cancelOrError,
                MamdaTradeRecap recap)
            {
                myOrigShortSaleCircuitBreaker = cancelOrError.getOrigShortSaleCircuitBreaker();
                Console.WriteLine("myOrigShortSaleCircuitBreaker: " + myOrigShortSaleCircuitBreaker);
            }

            public void onTradeCorrection(
                MamdaSubscription sub,
                MamdaTradeListener listener,
                MamaMsg msg,
                MamdaTradeCorrection correction,
                MamdaTradeRecap recap)
            {
                myCorrShortSaleCircuitBreaker = correction.getCorrShortSaleCircuitBreaker();
                Console.WriteLine("myCorrShortSaleCircuitBreaker: " + myCorrShortSaleCircuitBreaker);
            }

            public void onTradeClosing(
                MamdaSubscription sub,
                MamdaTradeListener listener,
                MamaMsg msg,
                MamdaTradeClosing closing,
                MamdaTradeRecap recap)
            {
            }

            public void onStale(
                MamdaSubscription subscription,
                mamaQuality quality)
            {
            }

            public void onError(
                MamdaSubscription subscription,
                MamdaErrorSeverity severity,
                MamdaErrorCode errorCode,
                String errorStr)
            {
            }
        }
    }
}